packages/grpc-js/src/server.ts

```diff
@@ -1421,41 +1420,72 @@ export class Server {
       }
 
       if (this.keepaliveTimeMs < KEEPALIVE_MAX_TIME_MS) {
-        keeapliveTimeTimer = setInterval(() => {
-          keepaliveTimeoutTimer = setTimeout(() => {
-            sessionClosedByServer = true;
-            session.close();
+        keepaliveInterval = setInterval(() => {
+          const keepaliveTimeout = setTimeout(() => {
+            if (keepaliveInterval) {
+              clearInterval(keepaliveInterval);
+              keepaliveInterval = null;
+              sessionClosedByServer = true;
+              this.trace('Connection dropped by keepalive timeout');
+              session.close();
+            }
           }, this.keepaliveTimeoutMs);
-          keepaliveTimeoutTimer.unref?.();
+          keepaliveTimeout.unref?.();
 
           try {
-            session.ping(
-              (err: Error | null, duration: number, payload: Buffer) => {
-                if (keepaliveTimeoutTimer) {
-                  clearTimeout(keepaliveTimeoutTimer);
+            if (
+              !session.ping(
```

Comment:
I don't like having a big statement in an `if` condition. I'd rather have a `pingSentSuccessfully` variable or something.